### PR TITLE
Restore trusted types

### LINKED ISSUE #1145: Unrevert Trusted Types Support
We had support for Trusted Types, but it was reverted prior to our most recent release along with some other security hooks that are likely to only be used internally at google.

The other hooks were too obscure, and their cost in bytes and runtime was so high that the benefit to the small number of people that might use it was outweighed by the cost. It still lives on in the security branch, which we will maintain.

However Trusted Types is now releasing in Chrome shortly, and has minimal size and runtime cost, while covering the most critical surface area.

### PR DESCRIPTION
**Pull request by rictic**
Re-adds trusted type support now that it's soon to land in Chrome stable

Size at master:
```
┌──────────────────────────────────────┐
│                                      │
│   Destination: lit-html.bundled.js   │
│   Bundle Size:  9.13 KB              │
│   Minified Size:  9.11 KB            │
│   Gzipped Size:  3.33 KB             │
│   Brotli size: 3.01 KB               │
│                                      │
└──────────────────────────────────────┘
created lit-html.bundled.js in 601ms
    3416
```

Size at c5199ab0f1f85fc93ce3bc2d48a409b63c9360a7:
```
lit-html.js → lit-html.bundled.js...
┌──────────────────────────────────────┐
│                                      │
│   Destination: lit-html.bundled.js   │
│   Bundle Size:  9.37 KB              │
│   Minified Size:  9.35 KB            │
│   Gzipped Size:  3.43 KB             │
│   Brotli size: 3.1 KB                │
│                                      │
└──────────────────────────────────────┘
created lit-html.bundled.js in 598ms
    3518
```

Fixes #1145

### CODE REVIEW

File: src/lib/parts.ts [hunk 1]
```diff
@@ -93,7 +115,12 @@ export class AttributeCommitter {
   commit(): void {
     if (this.dirty) {
       this.dirty = false;
-      this.element.setAttribute(this.name, this._getValue() as string);
+      let value = this._getValue() as string;
+      if (typeof value === 'symbol') {
```
Comment (justinfagnani) [inline src/lib/parts.ts:119]: moving the check for symbol into _getValue() might save a couple of bytes

File: src/lib/parts.ts [hunk 2]
```diff
@@ -69,11 +69,33 @@ export class AttributeCommitter {
...
+    // String(value)
+    // The exception is if v is an array, in which case we do want to smash
+    // it together into a string without calling String() on the array.
+    //
+    // This also allows trusted values (when using TrustedTypes) being
+    // assigned to DOM sinks without being stringified in the process.
+    if (l === 1 && strings[0] === '' && strings[1] === '' &&
+        parts[0] !== undefined) {
+      const v = parts[0].value;
+      if (!isIterable(v)) {
```
Comment (justinfagnani) [inline src/lib/parts.ts:90]: strings are iterable, so maybe check that too to avoid the loop below?

File: src/lib/parts.ts [hunk 3]
```diff
@@ -69,11 +69,33 @@ export class AttributeCommitter {
...
+    // string. Instead we want to just return the value itself directly,
+    // so that sanitizeDOMValue can get the actual value rather than
+    // String(value)
+    // The exception is if v is an array, in which case we do want to smash
+    // it together into a string without calling String() on the array.
+    //
+    // This also allows trusted values (when using TrustedTypes) being
+    // assigned to DOM sinks without being stringified in the process.
+    if (l === 1 && strings[0] === '' && strings[1] === '' &&
+        parts[0] !== undefined) {
```
Comment (justinfagnani) [inline src/lib/parts.ts:88]: Every part in the parts array should be defined

File: src/lib/template-result.ts [hunk 4]
```diff
@@ -100,7 +110,17 @@ export class TemplateResult {
 
   getTemplateElement(): HTMLTemplateElement {
     const template = document.createElement('template');
-    template.innerHTML = this.getHTML();
+    let value = this.getHTML();
+    const trustedTypes = window.trustedTypes;
+    if (trustedTypes && !policy) {
+      policy = trustedTypes.createPolicy('lit-html', {createHTML: (s) => s});
```
Comment (justinfagnani) [inline src/lib/template-result.ts:116]: Is creating a policy expensive at all? Why not move this into the `policy` initializer and assume it'll be used very shortly. A few less bytes and one less conditional in a perf-sensitive path.

File: src/lib/parts.ts (hunk 3 above)
Comment (rictic) [reply src/lib/parts.ts]: Removed check

File: src/lib/parts.ts (hunk 2 above)
Comment (rictic) [reply src/lib/parts.ts]: Added strings check

File: src/lib/parts.ts (hunk 1 above)
Comment (rictic) [reply src/lib/parts.ts]: Done

File: src/lib/template-result.ts (hunk 4 above)
Comment (rictic) [reply src/lib/template-result.ts]: Moved up and made policy a const, as a hint to the VM that the check for policy being defined can be removed, it's either always defined or always undefined.

File: src/lib/template-result.ts [hunk 5]
```diff
@@ -20,6 +20,21 @@ import {reparentNodes} from './dom.js';
...
+ * That HTML is a developer-authored constant, and is parsed with innerHTML
+ * before any untrusted expressions have been mixed in. Therefor it is
+ * considered safe by construction.
+ */
+const policy = ((): Pick<TrustedTypePolicy, 'createHTML'>|void => {
+  const tt = window.trustedTypes;
+  if (tt !== undefined) {
+    return tt.createPolicy('lit-html', {createHTML: (s) => s});
+  }
+})();
```
Comment (justinfagnani) [inline src/lib/template-result.ts:36]: Consider:

```ts
const policy = trustedTypes && trustedTypes.createPolicy('lit-html', {createHTML: (s) => s});
```

Comment (rictic) [reply src/lib/template-result.ts]: I like it. Saves another 14 bytes after gzip

Comment (rictic) [reply src/lib/template-result.ts]: Er, 13 bytes.

Adding `window.` adds only one byte to the gzipped output! It's a common enough substring, but that's still impressive!

Comment (justinfagnani) [reply src/lib/template-result.ts]: nice :)

**Comment by rictic** [comment]
Brendan released wct-mocha 1.1.0! So now we can run tests with trusted types enabled and disabled

**Comment by rictic** [comment]
Ignore the failing travis run, it's trying to pull from the other PR (now merged with this one) directly, which fails because the other PR is now closed

**Pull request merged**